src/main/java/org/embulk/input/google_ads/GoogleAdsReporter.java

```diff
@@ -53,40 +57,44 @@ private UserCredentials buildCredential(PluginTask task)
                 .build();
     }
 
-    public Iterable<GoogleAdsServiceClient.SearchPage> getReportPage()
-    {
-        List<GoogleAdsServiceClient.SearchPage> pages = new ArrayList<GoogleAdsServiceClient.SearchPage>();
-
-        String startDateTime = null;
-       do {
-            String query = buildQuery(task, startDateTime);
-            logger.info(query);
-            SearchGoogleAdsRequest request = buildRequest(task, query);
-            GoogleAdsServiceClient googleAdsService = client.getVersion14().createGoogleAdsServiceClient();
-            GoogleAdsServiceClient.SearchPagedResponse response = googleAdsService.search(request);
-
-            if (response.getPage().getResponse().getResultsCount() == 0) {
-                return pages;
-            }
+    private Iterable<GoogleAdsServiceClient.SearchPage> search(Map<String, String> params) {
+        String query = buildQuery(task, params);
+        logger.info(query);
+        SearchGoogleAdsRequest request = buildRequest(task, query);
+        GoogleAdsServiceClient googleAdsService = client.getVersion14().createGoogleAdsServiceClient();
+        GoogleAdsServiceClient.SearchPagedResponse response = googleAdsService.search(request);
+        return response.iteratePages();
+    }
 
-            response.iteratePages().iterator().forEachRemaining(pages::add);
+    public void search(Consumer<Iterable<GoogleAdsServiceClient.SearchPage>> consumer, Map<String, String> params) {
+        Iterable<GoogleAdsServiceClient.SearchPage> pages = search(params);
+        consumer.accept(pages);
 
-            if (task.getResourceType().equals("change_event")) {
-                GoogleAdsServiceClient.SearchPage lastPage = pages.get(pages.size() - 1);
-                GoogleAdsRow lastRow = null;
-                for(GoogleAdsRow row : lastPage.getValues()) {
-                    lastRow = row;
-                }
+        if (task.getResourceType().equals("change_event")) {
+            // reset iterator
+            pages.iterator();
+            GoogleAdsRow lastRow = fetchLastRow(pages);
+            if (lastRow == null) return ;
 
-                if (lastRow == null) {
-                    break;
-                } else {
-                    startDateTime = lastRow.getChangeEvent().getChangeDateTime();
-                }
-            }
-       } while (startDateTime != null && !startDateTime.isEmpty());
+            lastRow.getChangeEvent().getChangeDateTime();
+            Map<String, String> nextParams = new HashMap<>();
+            nextParams.put("start_datetime", lastRow.getChangeEvent().getChangeDateTime());
+            search(consumer, nextParams);
```

Comment:
There is a risk that pages may not be released.

Reply:
fixed: 003f153

I fixed to pass single `SearchPage` to consumer, and `pages`  is removed.